### PR TITLE
Add docstrings for applying model part

### PR DESCRIPTION
**Pull request by averywpx**

### CODE REVIEW

File: constraint_utils/constraint_basics.py [hunk 1]
```diff
@@ -54,11 +54,33 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
 
 
 def load_tx_expression_data(context=True):
+    """
+    Load tx MatrixTable with processed expression data.
+
+    :param context: Whether the tx Table will be used by context Table. defaults to True.
+    :return: tx Table with expression data.
+    """
     tx_ht_file = 'gs://gnomad-public/papers/2019-tx-annotation/pre_computed/all.possible.snvs.tx_annotated.021819.ht' if context \
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py:63]: Since there are only standard model and worst_csq model these two options as for now, this function may not be used

File: constraint_utils/constraint_basics.py [hunk 2]
```diff
@@ -54,11 +54,33 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
 
 
 def load_tx_expression_data(context=True):
+    """
+    Load tx MatrixTable with processed expression data.
+
+    :param context: Whether the tx Table will be used by context Table. defaults to True.
+    :return: tx Table with expression data.
+    """
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py:65]: Since there are only standard model and worst_csq model these two options as for now, this function may not be used.

File: constraint_utils/constraint_basics.py [hunk 3]
```diff
@@ -416,6 +444,13 @@ def add_most_severe_csq_to_tc_within_ht(t):
 
 
 def take_one_annotation_from_tc_within_ht(t):
+    """
+    Annotate 'transcript_consequences' using the 'transcript_consequences' on 0 index within the 
+    vep annotation of the input Table.
+
+    :param t: Input Table.
+    :return: Input Table with only one annotation for 'transcript_consequences'.
+    """
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py:456]: Since there are only standard model and worst_csq model these two options as for now, this function may not be used.

File: constraint_utils/constraint_basics.py [hunk 4]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
                             custom_model: str = None, dataset: str = 'gnomad',
                             impose_high_af_cutoff_upfront: bool = True, half_cutoff = False) -> hl.Table:
-
+    """
+    Compute the expected number of variants using plateau models and coverage model.
+
+    This function sums the number of possible variants times the mutation rate for all variants, and apply the calibration
+    model separately for CpG transitions and other sites. For sites with coverage lower than coverage cutoff, the value got 
+    from previous step is multiplied by the coverage correction factor. These values are summed across the set of variants 
+    of interest to obtain the expected number of variants.
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py:472]: I felt this description is too wordy, so I added a brief version below. However, the brief one is not that formal, so I wonder which one is better.

File: constraint_utils/constraint_basics.py [hunk 5]
```diff
@@ -635,9 +709,23 @@ def collapse_lof_ht(lof_ht: hl.Table, keys: Tuple[str], calculate_pop_pLI: bool
...
+        - annotation - could be 'most_severe_consequence' of either 'worst_csq_by_gene' or 'transcript_consequences', or 'csq' of 'tx_annotation'
+        - modifier - classic lof annotation, LOFTEE annotation, or PolyPhen annotation
+        - gene
+        - coverage
+        - expressed (optional)
+        - transcript
+        - canonical
+    
+    ..note::
+        HT must be exploded against whatever axis.
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py:727]: My understanding is that this corresponds to L514, L521, L535, so I kept this as a note. Correct me if I'm wrong

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (jkgoodrich) [inline .vscode/settings.json]: I'm assuming you didn't mean to include this file

File: constraint_utils/constraint_basics.py (hunk 2 above)
Comment (jkgoodrich) [reply constraint_utils/constraint_basics.py:65]: Yeah, so these are files created by Beryl for her transcript expression-aware annotation paper: https://github.com/macarthur-lab/tx_annotation

The context one is the path to a Table with proportion expressed across transcripts (pext) score annotations for all possible bases, and the other one is pext scores for sites found in gnomAD v2.1.1 exomes

File: constraint_utils/constraint_basics.py [hunk 7]
```diff
@@ -54,11 +54,33 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
...
+            - ensg
+            - csq
+            - symbol
+            - lof
+            - mean_expression
+            - max_expression
+            - mean_brain_expression
+            - Brain_Cortex
+
+        :param csq_expression: 'tx_annotation' within tx Table
```
Comment (jkgoodrich) [inline constraint_utils/constraint_basics.py]: it's not really the 'tx_annotation', it's an element of the 'tx_annotation' which is an array that has `.map` used on it. It's probably a struct.

File: constraint_utils/constraint_basics.py [hunk 8]
```diff
@@ -75,6 +97,12 @@ def process_expression_data(csq_expression):
 
 
 def annotate_distance_to_splice(input_ht):
+    """
+    Annotate nearest distance to splice site.
```
Comment (jkgoodrich) [inline constraint_utils/constraint_basics.py]: Do you mean "Annotate distance to the nearest splice site."?

File: constraint_utils/constraint_basics.py [hunk 9]
```diff
@@ -75,6 +97,12 @@ def process_expression_data(csq_expression):
 
 
 def annotate_distance_to_splice(input_ht):
+    """
```
Comment (jkgoodrich) [inline constraint_utils/constraint_basics.py:103]: Does this function actually work?

File: constraint_utils/constraint_basics.py [hunk 10]
```diff
@@ -416,6 +444,13 @@ def add_most_severe_csq_to_tc_within_ht(t):
 
 
 def take_one_annotation_from_tc_within_ht(t):
+    """
+    Annotate 'transcript_consequences' using the 'transcript_consequences' on the 0 index within the
```
Comment (jkgoodrich) [inline constraint_utils/constraint_basics.py]: Sorry, I think the summary description should be a single line and then you can add more detail on specifics below it

File: constraint_utils/constraint_basics.py [hunk 11]
```diff
@@ -416,6 +444,13 @@ def add_most_severe_csq_to_tc_within_ht(t):
 
 
 def take_one_annotation_from_tc_within_ht(t):
+    """
+    Annotate 'transcript_consequences' using the 'transcript_consequences' on the 0 index within the 
+    vep annotation of the input Table.
+
+    :param t: Input Table.
```
Comment (jkgoodrich) [inline constraint_utils/constraint_basics.py]: It can take a Table or a MatrixTable

File: .gitignore [hunk 12]
```diff
@@ -106,3 +106,6 @@ ENV/
 
 # tokens
 slack_creds.py
+
+# else
```
Comment (klaricch) [inline .gitignore]: why "else"?

File: constraint_utils/constraint_basics.py [hunk 13]
```diff
@@ -54,11 +54,33 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
...
+    Load tx MatrixTable with processed expression data.
+
+    :param context: Whether the tx Table will be used by context Table. defaults to True.
+    :return: tx Table with expression data.
+    """
     tx_ht_file = 'gs://gnomad-public/papers/2019-tx-annotation/pre_computed/all.possible.snvs.tx_annotated.021819.ht' if context \
         else 'gs://gnomad-public/papers/2019-tx-annotation/gnomad.exomes.r2.1.1.sites.tx_annotated.021319.ht'
     tx_ht = hl.read_matrix_table(tx_ht_file).rows()
 
     def process_expression_data(csq_expression):
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: add type for csq_expression and output

File: constraint_utils/constraint_basics.py (hunk 7 above)
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: maybe "Name of expression annotation ('tx_annotation' ) within tx Table."?

File: constraint_utils/constraint_basics.py [hunk 14]
```diff
@@ -54,11 +54,33 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
 
 
 def load_tx_expression_data(context=True):
+    """
+    Load tx MatrixTable with processed expression data.
+
+    :param context: Whether the tx Table will be used by context Table. defaults to True.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: what is meant by "will be used by context Table"? It looks like context means to load the Table with all possible SNVs?

File: constraint_utils/constraint_basics.py [hunk 15]
```diff
@@ -75,6 +97,12 @@ def process_expression_data(csq_expression):
 
 
 def annotate_distance_to_splice(input_ht):
+    """
+    Annotate nearest distance to splice site.
+
+    :param input_ht: Input exome or context Table.
+    :return: Input Table with 'nearest_splice' annotation.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :return: Table with 'nearest_splice' annotation added (distance to nearest splice site).
```

File: constraint_utils/constraint_basics.py [hunk 16]
```diff
@@ -75,6 +97,12 @@ def process_expression_data(csq_expression):
 
 
 def annotate_distance_to_splice(input_ht):
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: add type and output on this line

File: constraint_utils/constraint_basics.py [hunk 17]
```diff
@@ -416,6 +444,13 @@ def add_most_severe_csq_to_tc_within_ht(t):
 
 
 def take_one_annotation_from_tc_within_ht(t):
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: add type and output

File: constraint_utils/constraint_basics.py (hunk 10 above)
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    Annotate 'transcript_consequences' using only the first consequence (index 0) of 'transcript_consequences' within the 
```

File: constraint_utils/constraint_basics.py (hunk 11 above)
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: can also be a matrix table

File: constraint_utils/constraint_basics.py [hunk 18]
```diff
@@ -416,6 +444,13 @@ def add_most_severe_csq_to_tc_within_ht(t):
...
+    """
+    Annotate 'transcript_consequences' using the 'transcript_consequences' on the 0 index within the 
+    vep annotation of the input Table.
+
+    :param t: Input Table.
+    :return: Input Table with only one annotation for 'transcript_consequences'.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: can also be a matrix table

Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :return: Input Table with only the first consequence for the 'transcript_consequences' annotation.
```

File: constraint_utils/constraint_basics.py [hunk 19]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+        - possible_variants (including those for each population)
+        - expected_variants (including those for each population)
+        - mu - sum(mu_snp * possible_variant * coverage_correction)
+        - obs_exp - observed:expected ratio
+        - constraint annotated by `annotate_constraint_groupings`
+
+    :param exome_ht: Exome site Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param context_ht: Context Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param mutation_ht: Mutation rate Table with 'mu_snp' field.
+    :param plateau_models: a linear model that clibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :param plateau_models: A linear model that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
```

File: constraint_utils/constraint_basics.py [hunk 20]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+    :param coverage_model: a linear model that clibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :param coverage_model: A linear model that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
```

File: constraint_utils/constraint_basics.py [hunk 21]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+    :param recompute_possible: Whether to use context Table to recompute the number of possible variants instead of using a precomputed intermediate Table if it exists. Defaults to False.
+    :param remove_from_denominator: Whether to remove allels with high frequency in context Table from the denominater, defaults to True
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :param remove_from_denominator: Whether to remove alleles with high frequency in context Table from the denominator, defaults to True
```

File: constraint_utils/constraint_basics.py (hunk 20 above)
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: maybe should mention it consists of slope and intercept?

File: constraint_utils/constraint_basics.py [hunk 22]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+
+    This function sums the number of possible variants times the mutation rate for all variants, and apply the calibration
+    model separately for CpG transitions and other sites. For sites with coverage lower than coverage cutoff, the value got 
+    from previous step is multiplied by the coverage correction factor. These values are summed across the set of variants 
+    of interest to obtain the expected number of variants.
+    
+    A brief view of how to get the expected number of variants:
+    mu_agg = the number of possible variants * the mutation rate (all variants)
+    adjusted_mutation_rate = sum(plateau model slop * mu_agg + plateau model intercept) (separately for CpG transitions and other sites)
+    if 0 <= coverage < coverage cutoff:
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    if 0 < coverage < coverage cutoff:
```

File: constraint_utils/constraint_basics.py (hunk 4 above)
Comment (klaricch) [reply constraint_utils/constraint_basics.py:472]: i think keeping both descriptions is fine, it helps to have the additional details

File: constraint_utils/constraint_basics.py [hunk 23]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
                             recompute_possible: bool = False, remove_from_denominator: bool = True,
                             custom_model: str = None, dataset: str = 'gnomad',
                             impose_high_af_cutoff_upfront: bool = True, half_cutoff = False) -> hl.Table:
-
+    """
+    Compute the expected number of variants using plateau models and coverage model.
+
+    This function sums the number of possible variants times the mutation rate for all variants, and apply the calibration
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    This function sums the number of possible variants times the mutation rate for all variants, and applies the calibration
```

File: constraint_utils/constraint_basics.py [hunk 24]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+    model separately for CpG transitions and other sites. For sites with coverage lower than coverage cutoff, the value got
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    model separately for CpG transitions and other sites. For sites with coverage lower than the coverage cutoff, the value obtained 
```

File: constraint_utils/constraint_basics.py [hunk 25]
```diff
@@ -425,7 +460,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
                             recompute_possible: bool = False, remove_from_denominator: bool = True,
                             custom_model: str = None, dataset: str = 'gnomad',
                             impose_high_af_cutoff_upfront: bool = True, half_cutoff = False) -> hl.Table:
-
+    """
+    Compute the expected number of variants using plateau models and coverage model.
+
+    This function sums the number of possible variants times the mutation rate for all variants, and apply the calibration
+    model separately for CpG transitions and other sites. For sites with coverage lower than coverage cutoff, the value got 
+    from previous step is multiplied by the coverage correction factor. These values are summed across the set of variants
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    from the previous step is multiplied by the coverage correction factor. These values are summed across the set of variants 
```

File: constraint_utils/constraint_basics.py (hunk 21 above)
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: it looks like this filters the context ht to the keep criteria if the site is found in the exome_ht, or if not in the exome ht, just keeps that line in the context ht?

File: constraint_utils/constraint_basics.py [hunk 26]
```diff
@@ -635,9 +709,23 @@ def collapse_lof_ht(lof_ht: hl.Table, keys: Tuple[str], calculate_pop_pLI: bool
...
+    Add constraint annotations.
+    
+    Function adds the following annotations:
+        - annotation - could be 'most_severe_consequence' of either 'worst_csq_by_gene' or 'transcript_consequences', or 'csq' of 'tx_annotation'
+        - modifier - classic lof annotation, LOFTEE annotation, or PolyPhen annotation
+        - gene
+        - coverage
+        - expressed (optional)
+        - transcript
+        - canonical
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: these bottom two are added if custom model isn't specified

File: constraint_utils/constraint_basics.py [hunk 27]
```diff
@@ -635,9 +709,23 @@ def collapse_lof_ht(lof_ht: hl.Table, keys: Tuple[str], calculate_pop_pLI: bool
...
+    
+    ..note::
+        HT must be exploded against whatever axis.
+
+    :param ht: Input Table
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: or matrix table

File: constraint_utils/constraint_basics.py [hunk 28]
```diff
@@ -635,9 +709,23 @@ def collapse_lof_ht(lof_ht: hl.Table, keys: Tuple[str], calculate_pop_pLI: bool
...
+    :param custom_model: The customized model (one of "standard" or "worst_csq" for now), defaults to None.
+    :return: A tuple of input Table with annotations and the names of columns annotated.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: or matrix table

Comment (klaricch) [reply constraint_utils/constraint_basics.py]: columns -> names of added annotations

**Review by klaricch — CHANGES_REQUESTED** [review]

**Review by jkgoodrich — COMMENTED** [review]
These were some comments that I had started for my review before Kristen looked over this, but I didn't finish the review so I hadn't submitted them.

File: constraint_utils/constraint_basics.py (hunk 9 above)
Comment (averywpx) [reply constraint_utils/constraint_basics.py:103]: Not really. I got `Error summary: HailException: array index out of bounds: index=0, length=0` after I replaced all the `last_locus[0]` in this function with `last_locus[0].locus`. It's only called when `custom_model` is specified as `distance_to_splice`,  and Konrad said this model is not necessary as for now.

File: constraint_utils/constraint_basics.py (hunk 21 above)
Comment (averywpx) [reply constraint_utils/constraint_basics.py]: It's kind of wordy to include all the information. Is there any way to rephrase it?

File: constraint_utils/constraint_basics.py [hunk 29]
```diff
@@ -425,7 +463,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+        - constraint annotated by `annotate_constraint_groupings`
+
+    :param exome_ht: Exome site Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param context_ht: Context Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param mutation_ht: Mutation rate Table with 'mu_snp' field.
+    :param plateau_models: A linear model (output of `build_plateau_models_pop`) that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
+    :param coverage_model: A linear model (output of `build_coverage_model`) that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
+    :param recompute_possible: Whether to use context Table to recompute the number of possible variants instead of using a precomputed intermediate Table if it exists. Defaults to False.
+    :param remove_from_denominator: Whether to remove alleles in context Table with coverage and allele count equal to 0, frequency larger than `af_cutoff`, and can't be found in `exome_ht`, defaults to True
```
Comment (averywpx) [inline constraint_utils/constraint_basics.py]: It's kind of wordy to include all the information. Is there any way to rephrase it?

File: constraint_utils/constraint_basics.py [hunk 30]
```diff
@@ -57,14 +57,15 @@ def load_tx_expression_data(context=True):
     """
     Load tx MatrixTable with processed expression data.
 
-    :param context: Whether the tx Table will be used by context Table. defaults to True.
+    :param context: Whether to load the Table with proportion expressed across transcripts (pext) score annotations
+    for all possible bases or the Table with pext scores for sites found in gnomAD v2.1.1 exomes. defaults to True.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    for all possible bases or the Table with pext scores for sites found in gnomAD v2.1.1 exomes. Defaults to True.
```

File: constraint_utils/constraint_basics.py [hunk 31]
```diff
@@ -118,8 +119,8 @@ def annotate_distance_to_splice(input_ht):
     last_locus = hl.scan.take(ht.row, 1, ordering=-ht.locus.global_position())
     ht.key_by(
         interval=hl.or_missing((hl.len(last_locus) > 0) &
-                               (last_locus[0].contig == ht.locus.contig),
-                               hl.interval(last_locus[0], ht.locus))
+                               (last_locus[0].locus.contig == ht.locus.contig),
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: this isn't working, right?

File: constraint_utils/constraint_basics.py [hunk 32]
```diff
@@ -443,13 +444,15 @@ def add_most_severe_csq_to_tc_within_ht(t):
     return t.annotate_rows(vep=annotation) if isinstance(t, hl.MatrixTable) else t.annotate(vep=annotation)
 
 
-def take_one_annotation_from_tc_within_ht(t):
+def take_one_annotation_from_tc_within_ht(t: Union[hl.Table, hl.MatrixTable]) -> Union[hl.Table, hl.MatrixTable]:
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py:449]: "Union" needs to be imported at the top of the script:

"from typing import Union"

File: constraint_utils/constraint_basics.py [hunk 33]
```diff
@@ -443,13 +444,15 @@ def add_most_severe_csq_to_tc_within_ht(t):
...
-    Annotate 'transcript_consequences' using the 'transcript_consequences' on the 0 index within the 
+    Mutate 'transcript_consequences' annotation in vep.
+    
+    Annotate 'transcript_consequences' using only the first consequence (index 0) of 'transcript_consequences' within the
     vep annotation of the input Table.
 
-    :param t: Input Table.
-    :return: Input Table with only one annotation for 'transcript_consequences'.
+    :param t: Input Table or MatrixTable.
+    :return: Input Table or MatrixTable with only the first consequence for the 'transcript_consequences' annotation.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :return: Table or MatrixTable with only the first consequence for the 'transcript_consequences' annotation.
```

File: constraint_utils/constraint_basics.py [hunk 34]
```diff
@@ -443,13 +444,15 @@ def add_most_severe_csq_to_tc_within_ht(t):
     return t.annotate_rows(vep=annotation) if isinstance(t, hl.MatrixTable) else t.annotate(vep=annotation)
 
 
-def take_one_annotation_from_tc_within_ht(t):
+def take_one_annotation_from_tc_within_ht(t: Union[hl.Table, hl.MatrixTable]) -> Union[hl.Table, hl.MatrixTable]: 
     """
-    Annotate 'transcript_consequences' using the 'transcript_consequences' on the 0 index within the 
+    Mutate 'transcript_consequences' annotation in vep.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: "Mutate" sounds too vague. I think you can just use the below line as the main description.

File: constraint_utils/constraint_basics.py (hunk 29 above)
Comment (klaricch) [reply constraint_utils/constraint_basics.py]: sites in the context ht will be kept if not found in `exome_ht`

could condense to something like this:
Whether to remove alleles in context Table if found in 'exome_ht' and is not a PASS variant with an allele count greater than 0

File: constraint_utils/constraint_basics.py [hunk 35]
```diff
@@ -716,16 +719,16 @@ def annotate_constraint_groupings(ht: Union[hl.Table, hl.MatrixTable],
...
+        - canonical (added when custom model isn't specified as worst_csq or tx_annotation)
     
     ..note::
         HT must be exploded against whatever axis.
 
-    :param ht: Input Table
+    :param ht: Input Table or MatrixTable.
     :param custom_model: The customized model (one of "standard" or "worst_csq" for now), defaults to None.
-    :return: A tuple of input Table with annotations and the names of columns annotated.
+    :return: A tuple of input Table or MatrixTable with annotations and the names of added annotations.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :return: A tuple of input Table or MatrixTable with grouping annotations added and the names of added annotations.
```

File: constraint_utils/constraint_basics.py [hunk 36]
```diff
@@ -54,11 +54,34 @@ def load_all_possible_summary(filtered: bool = True) -> Dict[hl.Struct, int]:
 
 
 def load_tx_expression_data(context=True):
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: add types and output

File: constraint_utils/constraint_basics.py [hunk 37]
```diff
@@ -425,7 +463,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+    The expected_variants are summed across the set of variants of interest to obtain the final expected number of variants.
+    
+    Function adds the following annotations:
+        - variant_count - annotated by `count_variants` function
+        - adjusted_mutation_rate (including those for each population) - mutation rate adjusted by plateau model
+        - possible_variants (including those for each population)
+        - expected_variants (including those for each population)
+        - mu - sum(mu_snp * possible_variant * coverage_correction)
+        - obs_exp - observed:expected ratio
+        - constraint annotated by `annotate_constraint_groupings`
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: what is the "constraint" annotation?

File: constraint_utils/constraint_basics.py [hunk 38]
```diff
@@ -425,7 +463,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+
+    :param exome_ht: Exome site Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param context_ht: Context Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
+    :param mutation_ht: Mutation rate Table with 'mu_snp' field.
+    :param plateau_models: A linear model (output of `build_plateau_models_pop`) that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
+    :param coverage_model: A linear model (output of `build_coverage_model`) that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: should mention these consist of slope and intercept

File: constraint_utils/constraint_basics.py [hunk 39]
```diff
@@ -425,7 +463,46 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
+    :param recompute_possible: Whether to use context Table to recompute the number of possible variants instead of using a precomputed intermediate Table if it exists. Defaults to False.
+    :param remove_from_denominator: Whether to remove alleles in context Table with coverage and allele count equal to 0, frequency larger than `af_cutoff`, and can't be found in `exome_ht`, defaults to True
+    :param custom_model: The customized model (one of "standard" or "worst_csq" for now), defaults to None.
+    :param dataset: Dataset to use when computing frequency index, defaults to 'gnomad'.
+    :param impose_high_af_cutoff_upfront: Whether to remove alleles with allele frequency larger than `af_cutoff` (0.001), defaults to True.
+    :param half_cutoff: Whether to use half of `HIGH_COVERAGE_CUTOFF` as coverage cutoff. Otherwise `HIGH_COVERAGE_CUTOFF` will be used, defaults to False.
+    :return: Table with the expected number of variants.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :return: Table with 'obs_exp' annotation added (ratio of observed number of variants to expected number of variants).
```

File: constraint_utils/constraint_basics.py [hunk 40]
```diff
@@ -635,9 +712,23 @@ def collapse_lof_ht(lof_ht: hl.Table, keys: Tuple[str], calculate_pop_pLI: bool
 def annotate_constraint_groupings(ht: Union[hl.Table, hl.MatrixTable],
                                   custom_model: str = None) -> Tuple[Union[hl.Table, hl.MatrixTable], List[str]]:
     """
-    HT must be exploded against whatever axis
-
-    Need to add `'coverage': ht.exome_coverage` here (which will get corrected out later)
+    Add constraint annotations.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]: something like add annotations to be used for groupings would be more descriptive

**Review by klaricch — CHANGES_REQUESTED** [review]

File: constraint_utils/constraint_basics.py (hunk 31 above)
Comment (averywpx) [reply constraint_utils/constraint_basics.py]: No. It will raise the out of index exception.

File: constraint_utils/constraint_basics.py (hunk 32 above)
Comment (averywpx) [reply constraint_utils/constraint_basics.py:449]: Is there a reason Konrad didn't to that? He didn't import Any, Tuple, Optional, List, Dict as well.

File: constraint_utils/constraint_basics.py (hunk 38 above)
Comment (averywpx) [reply constraint_utils/constraint_basics.py]: I don't think so. The docstrings of `build_plateau_models_pop` and `build_coverage_model` already mentioned that the models are made of slopes and intercepts.

File: constraint_utils/constraint_basics.py (hunk 32 above)
Comment (klaricch) [reply constraint_utils/constraint_basics.py:449]: Union, Any, Tuple, etc. are mostly things that you have now added in the docstrings, so they either weren't used before or were potentially imported with "*"

File: constraint_utils/constraint_basics.py (hunk 38 above)
Comment (klaricch) [reply constraint_utils/constraint_basics.py]: it's unclear looking at this function what the Tuple part is of "plateau_models: Dict[str, Tuple[float, float]]" (is it slope then intercept, or intercept then slope?), it's helpful to specify exactly what is needed so that users don't have to jump around to other functions to figure it out

**Review by klaricch — CHANGES_REQUESTED** [review]

File: constraint_utils/constraint_basics.py [hunk 41]
```diff
@@ -491,8 +493,8 @@ def get_proportion_observed(exome_ht: hl.Table, context_ht: hl.Table, mutation_h
...
     :param context_ht: Context Table (output of `prepare_ht`) filtered to autosomes and pseudoautosomal regions.
     :param mutation_ht: Mutation rate Table with 'mu_snp' field.
-    :param plateau_models: A linear model (output of `build_plateau_models_pop`) that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
-    :param coverage_model: A linear model (output of `build_coverage_model`) that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
+    :param plateau_models: Linear models (output of `build_plateau_models_pop` with intercepts and slops) that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
+    :param coverage_model: A linear model (output of `build_coverage_model` with a intercept and a slop) that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
```
Comment (klaricch) [inline constraint_utils/constraint_basics.py]:
```suggestion
    :param plateau_models: Linear models (output of `build_plateau_models_pop`), with the values of the dictionary formatted as a Tuple of intercept and slope, that calibrates mutation rate to proportion observed for high coverage exome. It includes models for CpG site, non-CpG site, and each population in `POPS`.
    :param coverage_model: A linear model (output of `build_coverage_model`), formatted as a Tuple of intercept and slope, that calibrates a given coverage level to observed:expected ratio. It's a correction factor for low coverage sites.
```

**Review by klaricch — CHANGES_REQUESTED** [review]

**Comment by klaricch** [comment]
back to you @averywpx for one small change

**Review by klaricch — APPROVED** [review]
